coderd/healthcheck/workspaceproxy.go

```diff
@@ -0,0 +1,137 @@
+	// CurrentVersion is the current server version.
+	// We pass this in to make it easier to test.
+	CurrentVersion string
+	// FetchWorkspaceProxies is a function that returns the available workspace proxies.
+	FetchWorkspaceProxies *func(context.Context) (codersdk.RegionsResponse[codersdk.WorkspaceProxy], error)
+	// UpdateProxyHealth is a function called when healthcheck is run.
+	// This would normally be ProxyHealth.ForceUpdate().
+	// We do this because if someone mashes the healthcheck refresh button
+	// they would expect up-to-date data.
+	UpdateProxyHealth *func(context.Context) error
+}
+
+// @typescript-generate WorkspaceProxyReport
+type WorkspaceProxyReport struct {
+	Healthy  bool            `json:"healthy"`
+	Severity health.Severity `json:"severity"`
+	Warnings []string        `json:"warnings"`
+	Error    *string         `json:"error"`
+
+	WorkspaceProxies codersdk.RegionsResponse[codersdk.WorkspaceProxy]
+}
+
+func (r *WorkspaceProxyReport) Run(ctx context.Context, opts *WorkspaceProxyReportOptions) {
+	r.Healthy = true
+	r.Severity = health.SeverityOK
+	r.Warnings = []string{}
+
+	if opts.FetchWorkspaceProxies == nil {
+		return
+	}
+	fetchWorkspaceProxiesFunc := *opts.FetchWorkspaceProxies
+
+	if opts.UpdateProxyHealth == nil {
+		err := "opts.UpdateProxyHealth must not be nil if opts.FetchWorkspaceProxies is not nil"
+		r.Error = ptr.Ref(err)
+		return
+	}
+	updateProxyHealthFunc := *opts.UpdateProxyHealth
+
+	// If this fails, just mark it as a warning. It is still updated in the background.
+	if err := updateProxyHealthFunc(ctx); err != nil {
+		r.Severity = health.SeverityWarning
+		r.Warnings = append(r.Warnings, xerrors.Errorf("update proxy health: %w", err).Error())
+		return
+	}
+
+	proxies, err := fetchWorkspaceProxiesFunc(ctx)
+	if err != nil {
+		r.Healthy = false
+		r.Severity = health.SeverityError
+		r.Error = ptr.Ref(err.Error())
+		return
+	}
+
+	r.WorkspaceProxies = proxies
+
+	var total, healthy int
+	for _, proxy := range r.WorkspaceProxies.Regions {
+		total++
+		if proxy.Healthy {
+			healthy++
+		}
+
+		if len(proxy.Status.Report.Errors) > 0 {
+			for _, err := range proxy.Status.Report.Errors {
+				r.appendError(xerrors.New(err))
+			}
+		}
+	}
+
+	r.Severity = calculateSeverity(total, healthy)
+	r.Healthy = r.Severity != health.SeverityError
+
+	// Versions _must_ match. Perform this check last. This will clobber any other severity.
+	for _, proxy := range r.WorkspaceProxies.Regions {
+		if vErr := checkVersion(proxy, opts.CurrentVersion); vErr != nil {
+			r.Healthy = false
+			r.Severity = health.SeverityError
+			r.appendError(vErr)
+		}
+	}
+}
+
+// appendError multierror-appends err onto r.Error.
+// We only have one error, so multiple errors need to be squashed in there.
+func (r *WorkspaceProxyReport) appendError(errs ...error) {
+	var prevErr error
+	if r.Error != nil {
+		prevErr = xerrors.New(*r.Error)
+	}
+	r.Error = ptr.Ref(multierror.Append(prevErr, errs...).Error())
```

Comment:
How about `errors.Join` (stdlib) instead of `multierror` (hashicorp), or is there a benefit to using 3rd party?

Edit: On second though, perhaps doing the same as for warnings would make sense here, i.e. `r.Errors = append(r.Errors, err)`? (Then the both behave similarly).

A method `Error() error { return errors.Join(r.Errors...) }` could then be used if a single error is needed.

Reply:
Oh snap, I forgot about `errors.Join`.

Regarding the second thought, I wanted to keep the same struct shape as all the others, where there's a single `Error *string`. If we're going to do what you suggest, I'd prefer to apply it equally to all of the healthcheck report structs for the sake of consistency. Maybe a follow-up PR?